listenbrainz/webserver/static/js/src/recommendations/RecommendationCard.tsx

```diff
@@ -1,36 +1,129 @@
+  currentFeedback: RecommendationFeedBack;
+  updateFeedback: (recordingMbid: string, rating: RecommendationFeedBack) => void;
   className?: string;
   isCurrentUser: Boolean;
   currentUser?: ListenBrainzUser;
+  newAlert: (
+    alertType: AlertType,
+    title: string,
+    message: string | JSX.Element
+  ) => void;
+};
+
+type RecommendationCardState = {
+  feedback: RecommendationFeedBack;
 };
 
 export default class RecommendationCard extends React.Component<
-  RecommendationCardProps
+  RecommendationCardProps,
+  RecommendationCardState
 > {
+  APIService: APIService;
   playRecommendation: (recommendation: Recommendation) => void;
 
   constructor(props: RecommendationCardProps) {
     super(props);
 
+    this.state = {
+      feedback: props.currentFeedback || "feedback_not_given",
+    };
+
+    this.APIService = new APIService(
+      props.apiUrl || `${window.location.origin}/1`
+    );
+
     this.playRecommendation = props.playRecommendation.bind(
       this,
       props.recommendation
     );
   }
 
+  componentDidUpdate(prevProps: RecommendationCardProps) {
+    const { currentFeedback } = this.props;
+    if (currentFeedback !== prevProps.currentFeedback) {
+      this.setState({ feedback: currentFeedback });
+    }
+  }
+
+  submitFeedback = async (rating: RecommendationFeedBack) => {
+    const { recommendation, currentUser, isCurrentUser, updateFeedback } = this.props;
+    if (isCurrentUser && currentUser?.auth_token) {
+      const recordingMBID = _get(
+        recommendation,
+        "track_metadata.additional_info.recording_mbid"
+      );
+      if (rating === "remove_feedback") {
+        try {
+          const status = await this.APIService.deleteRecommendationFeedback(
+            currentUser.auth_token,
+            recordingMBID,
+          );
+          if (status === 200) {
+            this.setState({ feedback: "feedback_not_given" });
+            updateFeedback(recordingMBID, "feedback_not_given");
+          }
+        } catch (error) {
+          this.handleError(`Error while deleting recommendation feedback - ${error.message}`);
+        }
+      } else {
+        try {
+          const status = await this.APIService.submitRecommendationFeedback(
+            currentUser.auth_token,
+            recordingMBID,
+            rating
+          );
+          if (status === 200) {
+            this.setState({ feedback: rating });
+            updateFeedback(recordingMBID, rating);
+          }
+        } catch (error) {
+          this.handleError(`Error while submitting recommendation feedback - ${error.message}`);
+        }
+      };
+    }
+  };
+
+  handleError = (error: string | Error, title?: string): void => {
+    const { newAlert } = this.props;
+    if (!error) {
+      return;
+    }
+    newAlert(
+      "danger",
+      title || "Error",
+      typeof error === "object" ? error.message : error
+    );
+  };
+
+  handleClick(e) {
+    e.stopPropagation();
```

Comment:
@MonkeyDo this does not stop the drop-down from closing on click

Reply:
I'll have to look into this

Reply:
thanks :)

Reply:
So for some reason I'm having a hard time to run the PR locally, but at first glance, the `e.stopPropagation();` bit should be inside the function called when clicking the buttons `inside` the modal (the `RecommendationControl` components), not on the ellipsis button.

So you'll have something like 
```
<RecommendationControl
  icon={faAngry}
  title="I never want to hear this again!"
  action={(event: React.SyntheticEvent) =>
    this.submitFeedback(
      event,
      feedback === "hate" ? "remove_feedback" : "hate"
    )
  }
  className={`${feedback === "hate" ? " angry" : ""}`}
/>
```

Note that we're receiving the onClick event from RecommendationControl in the bound function (`action={(event: React.SyntheticEvent) => …`) and passing it to submitFeedback as the first argument (` this.submitFeedback(event,…`), so we need to modify submitFeedback like this:

```
submitFeedback = async (
    event: React.SyntheticEvent,
    rating: RecommendationFeedBack
  ) => {
    /* Stop click propagation so that the Bootstrap dropdown does not close on feedback click */
    event.stopPropagation();
    […rest of the function]
```

and the `action` in RecommendationControlProps type like this:
`action?: (event: React.SyntheticEvent) => void;`

Reply:
Can you tell me what issue you faced when trying to run it locally?

Reply:
and the dropdown still closes :(

Reply:
As decided on IRC, we don't want this feature now.